File: benchmarks/java/SecurityBenchmark.java
// Security Benchmark - Java
// Tests: bounds checking, null safety, overflow detection, memory safety

import java.util.Arrays;

public class SecurityBenchmark {

    // ─── Bounds Checking ────────────────────────────────────────────────────

    static double benchmarkBoundsChecking(int n) {
        long[] data = new long[n];
        for (int i = 0; i < n; i++) data[i] = i;

        long start = System.nanoTime();
        long sum = 0;
        for (int i = 0; i < n; i++) {
            // Bounds-checked access — throws ArrayIndexOutOfBoundsException
            sum += data[i];
        }
        double elapsed = (System.nanoTime() - start) / 1000.0;

        System.out.printf("Bounds-Checked Array Access (%d elements): %.2f µs%n", n, elapsed);
        return elapsed;
    }

    // ─── Null Safety (manual — no Option<T>) ────────────────────────────────

    static int findValue(int[] data, int target) {
        for (int i = 0; i < data.length; i++) {
            if (data[i] == target) return i;
        }
        return -1;  // Manual sentinel for "not found" — error-prone
    }

    static double benchmarkNullSafety(int n) {
        int[] data = new int[n];
        for (int i = 0; i < n; i++) data[i] = i;

        long start = System.nanoTime();
        long found = 0;
        for (int target = 0; target < n; target++) {
            if (findValue(data, target) >= 0) found++;
            if (findValue(data, n + target) < 0) found++;
        }
        double elapsed = (System.nanoTime() - start) / 1000.0;

        System.out.printf("Null Safety (sentinel) Searches (%d lookups): %.2f µs%n", n * 2, elapsed);
        return elapsed;
    }

    // ─── Integer Overflow Detection (manual) ────────────────────────────────

    static Long safeAdd(long a, long b) {
        long result = a + b;
        // Manual overflow check — easy to forget!
        if (((a ^ result) & (b ^ result)) < 0) {
            return null;  // Overflow detected (manual)
        }
        return result;
    }

    static double benchmarkOverflowDetection(int n) {
        long start = System.nanoTime();
        long count = 0;
        for (int i = 0; i < n; i++) {
            if (safeAdd(i, i) != null) count++;
            if (safeAdd(Long.MAX_VALUE - i, i + 1) == null) count++;
        }
        double elapsed = (System.nanoTime() - start) / 1000.0;

        System.out.printf("Overflow Detection (%d checks): %.2f µs%n", n * 2, elapsed);
        return elapsed;
    }

    // ─── String Safety ──────────────────────────────────────────────────────

    static double benchmarkStringSafety(int iterations) {
        long start = System.nanoTime();
        int totalLen = 0;
        for (int i = 0; i < iterations; i++) {
            // Java strings are UTF-16 — safe from buffer overflows
            String s = String.format("Item %d: 安全なUTF-8文字列 🔒🛡️", i);
            totalLen += s.length();
            // charAt is bounds-checked
            if (s.length() > 0) {
                char last = s.charAt(s.length() - 1);
            }
        }
        double elapsed = (System.nanoTime() - start) / 1000.0;

        System.out.printf("String Safety (%d UTF ops): %.2f µs%n", iterations, elapsed);
        return elapsed;
    }

    // ─── Memory Safety (GC-managed — no use-after-free) ─────────────────────

    static double benchmarkMemorySafety(int iterations) {
        long start = System.nanoTime();
        for (int i = 0; i < iterations; i++) {
            // GC prevents use-after-free — but adds pause overhead
            long[] data = new long[100];
            Arrays.fill(data, (long) i);
            long sum = 0;
            for (long v : data) sum += v;
        }
        double elapsed = (System.nanoTime() - start) / 1000.0;

        System.out.printf("Memory Safety (GC alloc/dealloc, %d iter): %.2f µs%n", iterations, elapsed);
        return elapsed;
    }

    // ─── Main ───────────────────────────────────────────────────────────────

    public static void main(String[] args) {
        System.out.println("=== Java Security Benchmark ===");
        System.out.println();
        System.out.println("Security analysis:");
        System.out.println("  ✓ Bounds checking on array access (runtime)");
        System.out.println("  ✗ Null safety — uses sentinel values, NullPointerExceptions possible");
        System.out.println("  ✗ Integer overflow — silent wraparound in Java (unchecked)");
        System.out.println("  ✓ UTF-16 safe string operations");
        System.out.println("  ~ Memory safety via GC (no use-after-free, but GC pauses)");
        System.out.println();

        double total = 0;

        total += benchmarkBoundsChecking(1_000_000);
        total += benchmarkNullSafety(10_000);
        total += benchmarkOverflowDetection(100_000);
        total += benchmarkStringSafety(100_000);
        total += benchmarkMemorySafety(100_000);

        System.out.println();
        System.out.printf("Total: %.2f µs (%.2f ms)%n", total, total / 1000.0);
        System.out.println();
        System.out.println("Security guarantees: PARTIAL (runtime checks only, no compile-time safety)");
    }
}

File: benchmarks/security_benchmark.rs
// Security Benchmark - Rust
// Tests: bounds checking, null safety, overflow detection, memory safety

use std::time::Instant;

// ─── Bounds Checking ────────────────────────────────────────────────────────

fn benchmark_bounds_checking(n: usize) -> f64 {
    let data: Vec<u64> = (0..n as u64).collect();
    let start = Instant::now();

    let mut sum = 0u64;
    for i in 0..n {
        // Bounds-checked access — panics on out-of-bounds instead of UB
        sum = sum.wrapping_add(data[i]);
    }
    // Prevent optimization from eliminating the loop
    std::hint::black_box(sum);

    let elapsed = start.elapsed().as_micros() as f64;
    println!("Bounds-Checked Array Access ({} elements): {:.2} µs", n, elapsed);
    elapsed
}

// ─── Null Safety (Option<T>) ────────────────────────────────────────────────

fn find_value(data: &[i32], target: i32) -> Option<usize> {
    for (i, &val) in data.iter().enumerate() {
        if val == target {
            return Some(i);
        }
    }
    None
}

fn benchmark_null_safety(n: usize) -> f64 {
    let data: Vec<i32> = (0..n as i32).collect();
    let start = Instant::now();

    // Search for existing and non-existing values
    let mut found = 0u64;
    for target in 0..n as i32 {
        if let Some(_idx) = find_value(&data, target) {
            found += 1;
        }
        // Search for value that doesn't exist
        if find_value(&data, n as i32 + target).is_none() {
            found += 1;
        }
    }
    std::hint::black_box(found);

    let elapsed = start.elapsed().as_micros() as f64;
    println!("Null Safety (Option) Searches ({} lookups): {:.2} µs", n * 2, elapsed);
    elapsed
}

// ─── Integer Overflow Detection ─────────────────────────────────────────────

fn safe_add(a: u32, b: u32) -> Option<u32> {
    a.checked_add(b)
}

fn benchmark_overflow_detection(n: u32) -> f64 {
    let start = Instant::now();

    let mut count = 0u64;
    for i in 0..n {
        if safe_add(i, i).is_some() {
            count += 1;
        }
        // This will overflow and return None (safe!)
        if safe_add(u32::MAX - i, i + 1).is_none() {
            count += 1;
        }
    }
    std::hint::black_box(count);

    let elapsed = start.elapsed().as_micros() as f64;
    println!("Overflow Detection ({} checks): {:.2} µs", n * 2, elapsed);
    elapsed
}

// ─── String Safety (UTF-8, no buffer overflows) ─────────────────────────────

fn benchmark_string_safety(iterations: usize) -> f64 {
    let start = Instant::now();

    let mut total_len = 0usize;
    for i in 0..iterations {
        // Safe string operations — no buffer overflows possible
        let s = format!("Item {}: 这是安全的UTF-8字符串 🔒🛡️", i);
        total_len += s.len();
        // Safe string slicing — panics on invalid UTF-8 boundaries
        let _safe_slice = &s[..s.char_indices().last().map(|(p, _)| p).unwrap_or(0)];
    }
    std::hint::black_box(total_len);

    let elapsed = start.elapsed().as_micros() as f64;
    println!("String Safety ({} UTF-8 ops): {:.2} µs", iterations, elapsed);
    elapsed
}

// ─── Memory Safety (No Use-After-Free, No Dangling Pointers) ────────────────

fn benchmark_memory_safety(iterations: usize) -> f64 {
    let start = Instant::now();

    for i in 0..iterations {
        // Ownership ensures no use-after-free
        let data = vec![i as u64; 100];
        let borrowed = &data;
        let sum: u64 = borrowed.iter().sum();
        // data is dropped here — borrows are already validated at compile time
        std::hint::black_box(sum);
    }

    let elapsed = start.elapsed().as_micros() as f64;
    println!("Memory Safety ({} alloc/dealloc): {:.2} µs", iterations, elapsed);
    elapsed
}

// ─── Main ───────────────────────────────────────────────────────────────────

fn main() {
    println!("=== Rust Security Benchmark ===");
    println!();
    println!("All operations are safe-by-default:");
    println!("  ✓ Bounds checking on array access");
    println!("  ✓ Null safety via Option<T> (no null pointers)");
    println!("  ✓ Integer overflow detection (checked arithmetic)");
    println!("  ✓ UTF-8 safe string operations");
    println!("  ✓ Ownership prevents use-after-free & dangling pointers");
    println!();

    let mut total = 0.0;

    total += benchmark_bounds_checking(1_000_000);
    total += benchmark_null_safety(10_000);
    total += benchmark_overflow_detection(100_000);
    total += benchmark_string_safety(100_000);
    total += benchmark_memory_safety(100_000);

    println!();
    println!("Total: {:.2} µs ({:.2} ms)", total, total / 1000.0);
    println!();
    println!("Security guarantees: ALL PASS (Rust enforces at compile time)");
}
